dlite_entities_service/models/soft.py

```diff
@@ -0,0 +1,148 @@
+"""Base (minimum set) models for SOFT entities."""
+from __future__ import annotations
+
+import difflib
+import re
+from typing import Annotated, Any
+
+from pydantic import (
+    AliasChoices,
+    BaseModel,
+    ConfigDict,
+    Field,
+    field_validator,
+    model_validator,
+)
+from pydantic.networks import AnyHttpUrl
+
+from dlite_entities_service.service.config import CONFIG
+
+URI_REGEX = re.compile(
+    r"^(?P<namespace>https?://.+)/(?P<version>\d(?:\.\d+){0,2})/(?P<name>[^/#?]+)$"
+)
+"""Regular expression to parse a SOFT entity URI."""
+
+
+class SOFTProperty(BaseModel):
+    """The minimum set of defining metadata for a SOFT Entity's property."""
+
+    model_config = ConfigDict(extra="forbid")
+
+    type_: Annotated[
+        str,
+        Field(
+            alias="type",
+            description="The type of the described property, e.g., an integer.",
+        ),
+    ]
+    shape: Annotated[
+        list[str] | None,
+        Field(
+            description=(
+                "The dimension of multi-dimensional properties. This is a list of "
+                "dimension expressions referring to the dimensions defined above. For "
+                "instance, if an entity have dimensions with names `H`, `K`, and `L` "
+                "and a property with shape `['K', 'H+1']`, the property of an instance "
+                "of this entity with dimension values `H=2`, `K=2`, `L=6` will have "
+                "shape `[2, 3]`."
+            ),
+            validation_alias=AliasChoices("dims", "shape"),
+        ),
+    ] = None
+    unit: Annotated[str | None, Field(description="The unit of the property.")] = None
+    description: Annotated[
+        str, Field(description="A human-readable description of the property.")
+    ]
+
+
+class SOFTEntity(BaseModel):
```

Comment:
In general the validators here seem quite complex, is there nothing built (either in python or pydantic ) that can replace them?

Reply:
Not that I know of? This is the way validators are done through pydantic. And this is the validation we need for ensuring the SOFT entities are actually ... well, valid :)